src/wallet/electrum.cpp

```diff
@@ -0,0 +1,237 @@
+ *
+ * This file is part of libbitcoin.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include <bitcoin/bitcoin/wallet/electrum.hpp>
+
+#include <algorithm>
+#include <cstdint>
+#include <boost/locale.hpp>
+#include <boost/range/adaptor/reversed.hpp>
+#include <bitcoin/bitcoin/define.hpp>
+#include <bitcoin/bitcoin/unicode/unicode.hpp>
+#include <bitcoin/bitcoin/unicode/unicode_ostream.hpp>
+#include <bitcoin/bitcoin/utility/assert.hpp>
+#include <bitcoin/bitcoin/utility/binary.hpp>
+#include <bitcoin/bitcoin/utility/collection.hpp>
+#include <bitcoin/bitcoin/utility/string.hpp>
+#include <bitcoin/bitcoin/utility/container_sink.hpp>
+#include <bitcoin/bitcoin/utility/ostream_writer.hpp>
+#include <bitcoin/bitcoin/wallet/dictionary.hpp>
+#include "../math/external/pkcs5_pbkdf2.h"
+
+namespace libbitcoin {
+namespace wallet {
+namespace electrum {
+
+#ifdef WITH_ICU
+
+// Electrum mnemonic private constants.
+static constexpr size_t hmac_iterations = 2048;
+static const auto hmac_data = to_chunk("Seed version");
+static const std::string passphrase_prefix = "electrum";
+
+
+static uint32_t special_modulo(int index_distance, int dictionary_length)
+{
+    if (index_distance < 0)
+        return dictionary_length - (-index_distance % dictionary_length);
+
+    return index_distance % dictionary_length;
+}
+
+static data_chunk old_mnemonic_decode(const word_list& mnemonic)
+{
+    static constexpr size_t mnemonic_word_multiple = 3;
+
+    if ((mnemonic.size() < mnemonic_word_multiple) ||
+        (mnemonic.size() % mnemonic_word_multiple) != 0)
+        return {};
+
+    const auto& lexicon = language::electrum::en_v1;
+    const auto dictionary_length = lexicon.size();
+    const auto seed_size = sizeof(uint32_t) * mnemonic.size() / 3;
+
+    data_chunk seed;
+    seed.reserve(seed_size);
+    data_sink ostream(seed);
+    ostream_writer sink(ostream);
+
+    auto all_found = false;
+    for(size_t i = 0; i < mnemonic.size() / 3; i += 3)
+    {
+        const auto index_1 = find_position(lexicon, mnemonic[i]);
+        const auto index_2 = find_position(lexicon, mnemonic[i+1]);
+        const auto index_3 = find_position(lexicon, mnemonic[i+2]);
+
+        all_found = ((index_1 != -1) && (index_2 != -1) && (index_3 != -1));
+        if (!all_found)
+            break;
```

Comment:
Recommend simply `return{}` here and elimination of the other code pertaining to `all_found`.